bindings/rust/s2n-tls/src/raw/callbacks.rs

```diff
@@ -0,0 +1,81 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::raw::{connection::Connection, enums::CallbackResult, error::Error};
+use core::{mem::ManuallyDrop, ptr::NonNull, task::Poll};
+use s2n_tls_sys::s2n_connection;
+
+const READY_OK: Poll<Result<(), Error>> = Poll::Ready(Ok(()));
+
+/// Convert the connection pointer provided to a callback into a Connection
+/// useable with the Rust bindings.
+///
+/// # Safety
+///
+/// This must ONLY be used for connection pointers provided to callbacks,
+/// which can be assumed to point to valid Connections because the
+/// callbacks were configured through the Rust bindings.
+pub unsafe fn with_connection<F, T>(conn_ptr: *mut s2n_connection, action: F) -> T
+where
+    F: FnOnce(&mut Connection) -> T,
+{
+    let raw = NonNull::new(conn_ptr).expect("connection should not be null");
+    let mut conn = Connection::from_raw(raw);
+    let r = action(&mut conn);
+    let _ = ManuallyDrop::new(conn);
+    r
+}
+
+pub(crate) trait AsyncCallback {
+    fn poll(&self, conn: &mut Connection) -> Poll<Result<(), Error>>;
+}
+
+pub(crate) trait AsyncCallbackExt {
+    fn trigger(self, conn: &mut Connection) -> CallbackResult;
+}
```

Comment:
it wasnt clear at first why this existed and how it was different from AsyncCallback. Can you add a comment explaining trigger vs poll.

- maybe rename `trigger`-> `set_callback`: the trigger and see if it completes is an internal detail.